crates/worker/src/partition/mod.rs

```diff
@@ -527,4 +548,44 @@ where
 
         Ok(is_duplicate)
     }
+
+    /// Tries to read as many records from the `log_reader` as are immediately available and stops
+    /// reading at `max_batching_size`.
+    async fn read_records<S>(
+        log_reader: &mut S,
+        max_batching_size: usize,
+        record_buffer: &mut Vec<(Lsn, Arc<Envelope>)>,
+    ) -> anyhow::Result<()>
+    where
+        S: Stream<Item = Result<anyhow::Result<(Lsn, Arc<Envelope>)>, restate_bifrost::Error>>
+            + Unpin,
+    {
+        // beyond this point we must not await; otherwise we are no longer cancellation safe
```

Comment:
Where can I read more to understand why this is?

Reply:
It's because on any await point we might get dropped if another select arm completes. So work that has happened as part of this future (e.g. consuming a record from Bifrost) would get lost.

https://users.rust-lang.org/t/cancel-safety-in-async-and-tokio-select/92381